reason-react-native/src/apis/AppRegistry.md

```diff
@@ -5,14 +5,103 @@ wip: true
 ---
 
 ```reason
+type task('data) = 'data => Js.Promise.t(unit);
+type taskProvider('data) = unit => task('data);
+type taskCanceller = unit => unit;
+type taskCancelProvider = unit => taskCanceller;
+
+type componentProvider('a) = unit => React.component('a);
+type wrapperComponentProvider('a, 'b) = 'b => React.component('a);
+
+type appParameters;
+
+external asAppParameters: 'a => appParameters = "%identity";
+
+type appConfig;
+
+[@bs.obj]
+external appConfig:
+  (
+    ~appKey: string,
+    ~component: componentProvider('a)=?,
+    ~run: appParameters => unit=?,
+    ~section: bool=?,
+    unit
+  ) =>
+  appConfig =
+  "";
+
+type runnable('a) = {
+  .
+  "component": Js.Nullable.t(componentProvider('a)),
+  "run": appParameters => unit,
+};
+
+type registry('a) = {
+  .
+  "sections": array(string),
+  "runnables": Js.Dict.t(runnable('a)),
+};
+
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external getAppKeys: unit => array(string) = "";
+
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external getRegistry: unit => registry('a) = "";
+
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external getRunnable: string => Js.Nullable.t(runnable('a)) = "";
+
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external getSectionKeys: unit => array(string) = "";
+
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external getSections: unit => Js.Dict.t(runnable('a)) = "";
+
+// multiple externals
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external registerComponent: (string, componentProvider('a)) => unit = "";
+
+// multiple externals
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external registerComponentWithSection:
+  (string, componentProvider('a), string) => unit =
+  "registerComponent";
+
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external registerConfig: array(appConfig) => unit = "";
+
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external registerRunnable: (string, appParameters => unit) => string = "";
+
 [@bs.module "react-native"] [@bs.scope "AppRegistry"]
-external registerComponent: (string, unit => React.component('a)) => unit =
+external registerSection: (string, componentProvider('a)) => unit = "";
+
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external runApplication: (string, 'a) => unit = "";
+
+[@bs.module "react-native"] [@bs.scope "AppRegistry"]
+external setWrapperComponentProvider:
+  Js.Nullable.t(wrapperComponentProvider('a, 'b)) => unit =
```

Comment:
Why is this nullable?

Reply:
`wrapperComponentProvider` is initiated with
```
let wrapperComponentProvider: ?WrapperComponentProvider;
```
I wasn't sure if setting that `null` might be a valid use pattern, so defined that as nullable.

On second though, it probably not needed at all.